ieducar/lib/Portabilis/Utils/Validation.php

```diff
@@ -95,4 +95,95 @@ public static function validatesCpf($cpf) {
 
     return ($primeiroDigito == $cpf[9] && $segundoDigito == $cpf[10]);
   }
+
+  /**
+   * Valida os novos formatos de Certidão Civil.
+   *
+   * @author Thieres Tembra <tdt@mytdt.com.br>
+   *
+   * @param  string  $certidao Certidão a ser validada.
+   * @param  bool  $motivo optional Se deseja retornar o motivo em string.
+   * @return bool|string Se !$motivo retorna true/false.
+   *                     Se $motivo retorna uma string com o motivo da falha ou uma string vazia (sucesso).
+   */
+  public static function validatesCertidaoNovoFormato($certidao, $motivo = false) {
```

Comment:
O que seria esse parâmetro `motivo`? Daria pra usar um nome mais autoexplicativo?

Reply:
Você leu a documentação da função? Linhas 105 a 107.
Acredito que não teremos um nome mais autoexplicativo que este. Talvez `$retMotivo` ?

Reply:
Algo como `$retornaMotivo` seria mais descritivo mas, analisando melhor talvez seja desnecessário esse parâmetro? Seria legal se o método retornasse apenas um tipo de dado: ou string ou booleano.

Pela aplicação do método no código, me parece que apenas o retorno da string é suficiente para indicar se houve ou não problema. Mas minha sugestão seria criar uma função de validação pra cada motivo e checar cada uma delas fazendo retornar apenas bool.

Reply:
Então, acredito que todas as validações dentro de `validatesCertidaoNovoFormato` deveriam ser funções separadas, cada uma desempenhando seu papel, dispensando comentários, pois o código precisa ser expressivo o suficiente pra não precisar de comentários. Sendo assim `validatesCertidaoNovoFormato` teria retorno booleano, você poderia fazer o use de `try catch` para os retornos dos motivos inves de uma variável para apontar a falha.

Reply:
@eberfreitas Fiz o método com dois tipos de retorno justamente por conta da sua aplicação. Se for interessante retornar o motivo do erro para o usuário estar ciente e corrigir, então o retorno é string. Entretanto como todo método de validação, o padrão é retornar booleano, por isso setei o valor default como `false`. Assim o utilizador precisa deixar claro se deseja o retorno do motivo do erro.

Eu poderia criar três funções novas, uma para cada motivo, mas isso aumentaria o trabalho na validação, pois  no momento de realizar a validação deste campo, _precisaria necessariamente lembrar_ que eu tenho três funções pra aplicar no mesmo campo, pois cada uma me retorna um erro diferente.

Esse também foi o motivo de utilizar como prefixo da função `validates` ao invés de `isValid`, pois no segundo caso obviamente fica claro que o retorno deve ser somente booleano.

Eu aceito todas as sugestões, mas como vocês perceberam discordo da necessidade de mudança em algumas delas. Portanto acredito que neste caso a melhor opção é a sugestão da @Bonot em utilizar `try`/`catch` para tratar estes erros, pois realmente resolve o problema e infelizmente nem me passou pela cabeça esta solução, por isso não a utilizei.

Dessa forma sugiro as seguintes definições:

- Tudo ficará em uma única função.
- O parâmetro `$motivo` será modificado para `$throwException` e também será booleano com valor padrão `false`.
- Apenas se o parâmetro `$throwException` for passado como `true` o código irá gerar uma nova `\Exception` com a mensagem de erro (motivo da não validação).
- O retorno passaria a ser somente booleano e não haveria necessidade de explicações na documentação do método.

Por favor me digam se estão de acordo para que eu possa prosseguir com as mesmas.

Reply:
Sobre:
>_Eu poderia criar três funções novas, uma para cada motivo, mas isso aumentaria o trabalho na validação, pois no momento de realizar a validação deste campo, precisaria necessariamente lembrar que eu tenho três funções pra aplicar no mesmo campo, pois cada uma me retorna um erro diferente._

Não sei se entendi bem o que você quis dizer, mas acho que a sugestão não foi validar 3 vezes e sim extrair os métodos e depois juntar tudo em uma única função. No caso criar funções separadas e depois usá-las dentro de uma única função. O que teoricamente atenderia à:
>_Tudo ficará em uma única função._

Quanto às outras sugestões, estou de acordo, considerando que, como o @munizeverton havia comentado, estamos buscando cada vez mais aplicar as boas práticas ao código, então acho importante sempre escrevermos códigos pensando em facilitar a interpretação e consequentemente manutenção do mesmo.

Reply:
Minha única ressalva é que as Exceptions devem ser lançadas sempre, sem necessidade de um parâmetro pra isso.

Reply:
@Bonot Se não for para utilizá-las independentemente, acho desnecessário termos três funções separadas e então utilizá-las dentro da função que será chamado para realizar a validação. A única validação complexa realmente é o cálculo do módulo 11. Repetindo mais uma vez: pra mim seria uma refatoração desnecessária. Entretanto se vocês acharem isto necessário, não me oponho em fazê-lo.

Quando citei o objetivo de ter uma única função era justamente para não ter que usar funções independentes para validar tamanho ou caracteres inválidos.

@eberfreitas Neste caso, o que aconteceria com o `return false` então? Pois ele nunca seria alcançado. Por isso sugiro a criação do parâmetro. O desenvolvedor pode querer apenas saber o retorno (se falhou ou não) sem querer o detalhe da falha. Se formos seguir por este caminho, não haverá portanto nem necessidade da função ter um retorno. Entretanto acredito que uma função de validação exige no mínimo um retorno booleano. O que você acha?

Fiquem tranquilos quanto as sugestões e críticas, tudo é bem vindo. E realmente quando desenvolvi esta função não me passou pela cabeça o uso de `try`/`catch` às 04h da manhã de um domingo pra segunda :sleeping:.

Concordo que o código deve ser melhorado sempre, inclusive me deu bastante agonia desenvolver no mesmo padrão dos outros arquivos que modifiquei, pois a indentação, por exemplo, não batia com o padrão definido no `.editorconfig` :disappointed:

Reply:
Então, foi sugerido fazer assim (fazendo funções para cada coisa) por conta da convenção do SOLID, que são princípios de boas práticas, uma das coisas que ele prega é que cada coisa deve ter uma unica responsabilidade, isso facilita a legibilidade e manutenção do código.

Sobre a questão de independência das funções, acho que até a parte de verificação de dígitos pode ser reaproveitada, da pra criar uma função que passa a string e a quantidade que deve ter e usar isso em outras partes, já que também tem que validar quantidade dos ineps.

Reply:
Estou ciente dos princípios do SOLID, e acredito que o conceito de _Single responsability_ foi alcançado pelo método.

Quanto a parte de verificação de dígitos ser reaproveitada, poderia ser, mas vamos lá é apenas a verificação de um `strlen`, isso não seria demais?

Assim irei fazer a refatoração gerando dois métodos:

1) Para verificar valores válidos em uma string, que receberá como primeiro parâmetro a string a ser verificado e segundo parâmetro uma string que deverá conter a expressão regular com valores válidos. Retornará booleano.
2) Para validar o módulo 11, que receberá como primeiro parâmetro a string a ser calculada o módulo 11, segundo parâmetro o dígito a ser comparado em string (para atender o caso do X), e terceiro parâmetro um array de strings com os valores válidos para caso o módulo 11 dê 10 (para atender o 1 ou X). Retornará booleano.

Em seguida irei refatorar tanto o método `validatesCertidaoNovoFormato` para atender as condições aqui discutidas, quanto o `validatesCpf` para usar os novos métodos criados.

Todos de acordo? Irei fazer as alterações agora a noite e já já faço o commit.

Reply:
>_Estou ciente dos princípios do SOLID, e acredito que o conceito de Single responsability foi alcançado pelo método._

Bem, não vejo como alcançado. O método possui muitas responsabilidades: validar quantidade de caracteres, passar para maiúsculo, pegar tantos digitos aqui, outros tantos lá, verificar se todos os valores são validos, etc.

>_Quanto a parte de verificação de dígitos ser reaproveitada, poderia ser, mas vamos lá é apenas a verificação de um strlen, isso não seria demais?_

É esse tipo de pensamento que estamos querendo evitar. Cada detalhe vale sim, e muito! Costumamos não nos importar com eles, mas é isso que faz virar uma sopa lá na frente.

Enfim, sabemos que o iEducar possui muitos gargalos e que tem poucos exemplos de boas práticas, mas estamos buscando manter o código limpo.
Peço desculpas se estamos nos estendendo demais, mas é importante que todos estejamos alinhados quanto a isso.

[Clean Code PHP](https://github.com/jupeter/clean-code-php#functions)

_PS: Seguirei com o code review do seu ultimo commit._

Reply:
@tembra parabéns pela sua contribuição e empenho em melhorar o código, é notório a evolução do mesmo desde o primeiro commit.

Conforme a @Bonot disse, hoje o iEducar está passando por uma transição quanto a melhoria de código, padrões, conceitos e metodologias. Os mesmos ainda não estão bem definidos e nem muito claros, por isso este PR foi muito comentado, pois é assim que iremos externar e discutir pontos de melhoria tanto internamente (core) quanto em comunidade.

Com o tempo vamos organizar e expor as boas práticas que queremos adotar assim como metodologias e padrões de desenvolvimento.

Atualmente, não contamos com testes unitários, mas este é um dos primeiros passos que teremos pela frente e queremos nos "doutrinar", de que o código pode sempre ser melhorado e refatorado a fim de ficar mais claro, coeso e desacoplado.

Abaixo, deixo minha contribuição de como poderia ser refatorado o seu código a fim de torná-lo mais simples e desacoplado. O código está longe de ser o ideal, a nomenclatura ainda não é a adequada, a utilização excessiva de métodos estáticos deve ser desencorajada, assim como outros detalhes.

```
/**
 * Valida se a $certidao contém 32 caracteres.
 *
 * @param string $certidao
 *
 * @throws \Exception
 *
 * @return string
 */
public static function validaTamanhoCertidaoNovoFormato($certidao)
{
    if (mb_strlen($certidao) === 32) {
        return $certidao;        
    }
    
    throw new \Exception("O tamanho da certidão '{$certidao}' é inválido");
}

/**
 * Valida se a parte numérica da $certidao contém apenas números.
 *
 * @param string $certidao
 *
 * @throws \Exception
 *
 * @return string
 */
public static function validaParteNumericaCertidaoNovoFormato($certidao)
{
    $parteNumerica = mb_substr($certidao, 0, 30);
    
    if (self::hasOnlyValidChars($parteNumerica, '/\d/')) {
        return $certidao;
    }
    
    throw new \Exception("A parte numérica da certidão '{$certidao}' contém caracteres inválidos");
}

/**
 * Valida se os dígitos verificadores da certidão possuem apenas números ou XX.
 *
 * @param string $certidao
 *
 * @throws \Exception
 *
 * @return string
 */
public static function validaCaracteresDigitosVerificadoresCertidaoNovoFormato($certidao)
{
    $digitosVerificadores = mb_substr($certidao, 30, 2);
    
    if (self::hasOnlyValidChars($digitosVerificadores, '/\d|X/')) {
        return $certidao;
    }
    
    throw new \Exception("Os dígitos verificadores da certidão '{$certidao}' contém caracteres inválidos");
}

/**
 * Valida os dígitos verificadores da certidão.
 *
 * @param string $certidao
 *
 * @throws \Exception
 *
 * @return string
 */
public static function validaDigitosVerificadoresCertidaoNovoFormato($certidao)
{
    $parteNumerica = mb_substr($certidao, 0, 30);
    $primeiroDigito = mb_substr($certidao, 30, 1);
    $segundoDigito = mb_substr($certidao, 31, 1);

    if (! static::isValidMod11Digit($parteNumerica, $primeiroDigito, '1')) {
        throw new \Exception("O módulo do primeiro dígito verificador da certidão '{$certidao}' é inválido");
    }
    
    if (! static::isValidMod11Digit($parteNumerica . $primeiroDigito, $segundoDigito, '1')) {
        throw new \Exception("O módulo do segundo dígito verificador da certidão '{$certidao}' é inválido");
    }
    
    return $certidao;
}

/**
 * Valida se a $certidao está no novo formato de Certidão Civil.
 *
 * @param string $certidao
 *
 * @throws \Exception Se $certidao for inválida
 *
 * @return string
 */
public static function validaCertidaoNovoFormato($certidao)
{
    $certidao = mb_strtoupper($certidao);
    
    static::validaTamanhoCertidaoNovoFormato($certidao);
    static::validaParteNumericaCertidaoNovoFormato($certidao);
    static::validaCaracteresDigitosVerificadoresCertidaoNovoFormato($certidao);
    
    // Verifica se o módulo 11 não foi calculado e portanto informado como XX 
    // conforme publicação do CNJ no Diário da Justiça - Edição nº 198/2009 
    // (Página 18) - Provimento nº 3, Artigo 7º, Inciso IX, Parágrafo 2º
    
    if (mb_substr($certidao, 30, 2) === 'XX') {
        return $certidao;
    }
    
    static::validaDigitosVerificadoresCertidaoNovoFormato($certidao);

    return $certidao;
}

/**
 * Indica se a $certidao está no novo formato de Certidão Civil.
 *
 * @param string $certidao
 *
 * @return boolean
 */
public static function isValidCertidaoNovoFormato($certidao)
{
    try {
        static::validatesCertidaoNovoFormato($certidao);
    } catch (Exception $e) {
        return false;
    }
    
    return true;
}
```

Este código, porém, se torna mais "testável" por ser desacoplado e seus métodos terem apenas uma responsabilidade. 

Este é o norte que queremos deixar a todos os contribuidores do projeto.

Espero ter contribuído um pouco com a evolução deste PR.

Desde já queremos agradecer muito a sua contribuição e principalmente sua compreensão das melhorias e discussões propostas neste PR. 

Abraço @tembra.

Reply:
Concordo com todos os pontos citados. Inclusive o exemplo de código do @edersoares é exatamente como eu faria se fosse construir uma classe do zero. E é aí que está o ponto. Tudo depende do contexto.

Originalmente em _OOP_, o princípio de _Single responsability_ foi criado para o contexto de classes, porém posteriormente fora estendido para os seus métodos. Quando citei que o método atendia este princípio, o fiz com a certeza de que o mesmo pertence a classe `Validation` e ele tinha uma única responsabilidade: validar a certidão civil.

Quando passamos a refatorar o código, percebemos que agora o que não segue mais o princípio de _Single responsability_ (já no meu último commit e ainda mais com o exemplo de código do @edersoares) é a classe `Validation`. Ou seja, existe neste momento a necessidade de se criar uma classe `Validation\CertidaoCivil`, para que possamos ter o nível de refatoração desejado, afinal temos mais de cinco métodos que terminam com `CertidaoNovoFormato`.

Se fosse fazê-lo, seria exatamente com os métodos criados pelo @edersoares com poucas mudanças, nas quais, por exemplo, a maioria dos métodos passariam a ter o escopo `private`. Neste caso @Bonot a simples comparação do tamanho de uma string faria completo sentido, não pelo fato de podermos reaproveitá-la (pois realmente não faz sentido!) e sim por dizer respeito a regra que a classe necessita para cumprir seu objetivo principal: validar a certidão civil.

Infelizmente, como já dito por vocês, o projeto do i-Educar ainda não possui um padrão de desenvolvimento a ser seguido. Por este motivo evitei criar novos arquivos em meu PR e me limitar em modificar somente os já existentes, pois acreditei que a análise do mesmo seria mais rápida.

No demais não se desculpem por expressarem sua opinião e ponto de vista. Isto é necessário para que a comunidade cresça cada vez mais e nós possamos conhecer uns aos outros :smiley:

Reply:
Depois do comentário do @edersoares não tenho mais nada a acrescentar.

>_Por este motivo evitei criar novos arquivos em meu PR e me limitar em modificar somente os já existentes, pois acreditei que a análise do mesmo seria mais rápida._

Entendo, mas fique a vontade para entregar o seu melhor :smiley:

Reply:
Excelente ponto de vista @tembra, após meu review a @Bonot em uma conversa interna, levantou o mesmo que você se referiu, quebrarmos em classes isoladas a fim de isolarmos de verdade a responsabilidade de cada uma!

Parabéns pelo seu conhecimento, é de pessoas assim que necessitamos no projeto.

E faço das palavras da @Bonot as minhas, caso você deseje, pode entregar seu melhor, dividindo as responsabilidades em classes isoladas. Será uma contribuição excepcional e um use case para futuras contribuições.

Abração @tembra!